cpp/algo_module/algorithm/algo.cpp

```diff
@@ -1 +1,204 @@
+    throw mgp::ValueException(
+        "Latitude and longitude properties, or a custom heuristic value, must be specified in every node!");
+  }
+  if (latitude_source.IsNumeric() && longitude_source.IsNumeric()) {
+    return haversineDistance(latitude_source.ValueNumeric(), longitude_source.ValueNumeric(), nodes.latLon.first,
+                             nodes.latLon.second);
+  }
+  throw mgp::ValueException("Latitude and longitude must be numeric data types!");
+}
+
+std::pair<double, double> Algo::TargetLatLon(const mgp::Node &target, const Config &config) {
+  if (config.heuristic_name != "") {  // if custom heuristic, dont return latitude and longitude
+    return std::make_pair<double, double>(0, 0);
+  }
+
+  auto latitude = target.GetProperty(config.latitude_name);
+  auto longitude = target.GetProperty(config.latitude_name);
+  if (latitude.IsNull() || longitude.IsNull()) {
+    throw mgp::ValueException("Latitude and longitude properties of the target node must be specified!");
+  }
+  if (latitude.IsNumeric() && longitude.IsNumeric()) {
+    return std::make_pair<double, double>(latitude.ValueNumeric(), longitude.ValueNumeric());
+  }
+  throw mgp::ValueException("Latitude and longitude must be numeric data types!");
+}
+
+double Algo::CalculateDistance(const Config &config, const mgp::Relationship &rel) {
+  if (config.unweighted) {  // return same distance if unweighted
+    return 10;
+  }
+  auto distance = rel.GetProperty(config.distance_prop);
+  if (distance.IsNull()) {
+    throw mgp::ValueException("If the graph is weighted, distance property of the relationship must be specified!");
+  }
+  if (distance.IsNumeric()) {
+    return distance.ValueNumeric();
+  }
+  if (distance.IsDuration() && config.duration) {
+    return distance.ValueDuration().Microseconds();
+  }
+  throw mgp::ValueException("Distance property must be a numeric or duration datatype!");
+}
+
+bool Algo::RelOk(const mgp::Relationship &rel, const Config &config,
+                 const bool in) {  // in true incoming, in false outgoing
+  if (config.in_rels.size() == 0 && config.out_rels.size() == 0) {
+    return true;
+  }
+
+  if (in && config.in_rels.find(std::string(rel.Type())) != config.in_rels.end()) {
+    return true;
+  }
+
+  if (!in && config.out_rels.find(std::string(rel.Type())) != config.out_rels.end()) {
+    return true;
+  }
+
+  return false;
+}
+
+bool Algo::LabelOk(const mgp::Node &node, const Config &config) {
+  bool whitelist_empty = config.whitelist.empty();
+
+  for (auto label : node.Labels()) {
+    if (config.blacklist.find(std::string(label)) != config.blacklist.end()) {
+      return false;
+    }
+    if (!whitelist_empty && config.whitelist.find(std::string(label)) == config.whitelist.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+void Algo::ParseRelationships(const std::shared_ptr<NodeObject> &prev, bool in, const GoalNodes &nodes, Lists &lists,
+                              const Config &config) {
```

Comment:
Should this be named parse relationships or expand relationships

Reply:
I mean it parses them, and if they are good expands them, so I dont know?

Reply:
Use expand, talked on meet